File: demo/demo_py/tools/demo01_record_write.py
#!/usr/bin/env python3

"""  
    需求：将 student 消息写入到磁盘文件
    实现：
        1.创建写出对象；
        2.设置写出参数；
        3.打开文件流；
        4.组织消息；
        5.写出消息；
        6.关闭文件流。


"""

from cyber.python.cyber_py3 import cyber
from cyber.python.cyber_py3 import record
from cyber.demo_base_proto.student_pb2 import Student

if __name__ == "__main__":
    cyber.init()
    print("record write ....")
    # 1.创建写出对象；
    writer = record.RecordWriter()
    # 2.设置写出参数；
    writer.set_size_fileseg(0)
    writer.set_intervaltime_fileseg(0)
    # 3.打开文件流；
    writer.open("/apollo/cyber/demo_py/tools/mytest_py.record")
    # 4.组织消息；
    stu = Student()
    stu.name = "haha"
    stu.height = 1.78
    stu.books.append("yuwen")
    stu.books.append("shuxue")
    stu.books.append("yingyu")
    print("full name:%s" %stu.DESCRIPTOR.full_name)
    channel = "chatter_py"
    writer.write_channel(channel,"cyber.demo_base_proto.Student","test....")
    for seq in range(20):
        stu.age = seq
        writer.write_message(channel,stu.SerializeToString(),seq)

    # 5.写出消息；
    # 6.关闭文件流。
    writer.close()

File: demo/demo_cc/tools/demo01_record_write.cc
/*

    将话题消息写入磁盘文件

*/
#include "cyber/demo_base_proto/student.pb.h"

#include "cyber/cyber.h"
#include "cyber/record/record_writer.h"
using apollo::cyber::demo_base_proto::Student;
using apollo::cyber::record::RecordWriter;

int main(int argc, char *argv[]) {
  apollo::cyber::Init(argv[0]);
  AINFO << "record 写操作演示...";
  // 2.创建 recorder 写对象；
  RecordWriter writer;
  // 3.设置写出参数；
  writer.SetSizeOfFileSegmentation(0);
  writer.SetIntervalOfFileSegmentation(0);
  // 4.打开文件流；
  writer.Open("/apollo/cyber/demo_cc/tools/mytest.record");
  AINFO << "file : " << writer.GetFile();

  // 5.写出消息话题以及消息数据；
  writer.WriteChannel("chatter", "apollo.cyber.demo_base_proto.Student",
                      "test ....");
  for (size_t i = 0; i < 10; i++) {
    auto stu = std::make_shared<Student>();
    stu->set_name("huluwa");
    stu->add_books("yingyu");
    stu->add_books("yuwen");
    // stu->SerializeAsString();
    std::string content;
    stu->SerializeToString(&content);
    writer.WriteMessage("chatter", content, i);
    // writer.WriteMessage(channel,stu->SerializeAsString(),i);
  }

  // 6.关闭文件流。
  writer.Close();

  return 0;
}